---
title: "conditional-independence-tests"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{conditional-independence-tests}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(pcgen2)
library(purrr)
```


## Using Willem's data: ------------------------------------------------------------

This is a dataset from Willem, composed of:
- `d`: a dataframe of 500 genotypes, 2 replicates from true graph G -> Y1 -> Y2
- `dm`: a dataframe of genotypic marker means and trait means across 100 SNPs that generate genetic effects

```{r}
library(Matrix)

setwd("~/github_repos/network-analysis/src/haplotypes/willem")
load(file = 'simdata1.RData')
str(d)
dm[1:6, 1:6]
dm[,4] %>% unique() # dm has 2 unique values for each snp 
                    # because data is simulated inbred plant

M <- as.matrix(dm[,-(1:3)] / sqrt(ncol(dm)))
M[1:6, 1:6]
K <- M %*% t(M)

str(K)

###############################################################
##  testing conditional independence, type (A) ############----
###############################################################

## Example of type (A): test if Y1 \independent G (S is empty)     
# Gives a positive genetic variance : G -> Y1 -> Y2
## Y1 ~ G ----------------------------------------------------

## rrBLUP library ---------------------------------------------
library(rrBLUP)
b <- kin.blup(data = dm, geno = 'G',pheno = "Y1",K = K)
b$Vg

## lme4qtl and lme4 libraries ---------------------------------
library(lme4qtl)
library(lme4)
library(RLRsim)
m <- relmatLmer(Y1 ~ (1|G) , # Write the formula of your model
                dm,  # Write your data table into a data frame
                relmat = list(G = K)) # Compute the relation 
                                      # matrix across grouping
                                      # IDs

# Estimating the heritibility and checking the variances:
# variance is congruent with above, and not zero
vf <- as.data.frame(VarCorr(m))[,c("grp", "vcov")]
vf$prop <- with(vf, vcov / sum(vcov))
vf

# Checking the residuals from models b and m are identical:
plot(residuals(m), b$resid)

# - extract the residual log-likelihood
# - also extract the residual log-likelihood for the model 
#   with sigma_G^2 = 0
# - compare LRT statistic to null distribution 
#   (.5/.5 mixture of point mass at zero and chi-square(1))

RLRsim::exactRLRT(m) # low p-value --> variance is not zero


# stats::confint(m) # the interval for sigma does not include zero, 
                  # significant random effect


### gaston ------------------------------------------------
library(gaston)
mg <- lmm.aireml(Y = dm$Y1 , K = K, verbose = F)
mg$tau  # The genetic variance, matches above

lrt <- 2*(mg$logL - mg$logL0) # similar lrt
pchisq(lrt, df = 1, lower.tail = F) 

## Another example of type (A): test if Y2 \independent G | Y1 
## Y2 ~ (1|G) + Y1, (ie S = {Y1}) ----------------------------
# should give b$Vg close to zero
# Probably, one should use a package that can exploit that
# K has rank lower than n = 500
# (lme4qtl ? pedigree ? breedR ? synbreed ? BGLR ? )

## rrBLUP library ---------------------------------------------
b2 <- kin.blup(data = dm, geno = 'G',pheno = "Y2", fixed = "Y1", 
               K = K)
b2$Vg  # no convergence

## lme4qtl and lme4 libraries ---------------------------------
m2 <- relmatLmer(Y2 ~ (1|G) + Y1 , # the formula of the model
           dm,      # Write your data table into a data frame
           relmat = list(G = K)) # the relation matrix 

vf2 <- as.data.frame(VarCorr(m2))[, c("grp", "vcov")]
vf2$prop <- with(vf2, vcov / sum(vcov))
vf2   # variance is zero as it should

# checking Residual likelihood
exactRLRT(m2) #p-value of 1 --> variance is zero 
# stats::confint(m2) # Not sure. sig01 is [0,.1] and sigma is [.4, .5],
                   # and intercept -s [-.06, .02]
                   # significant random effect?


### gaston ------------------------------------------------
mg2 <- lmm.aireml(Y = dm$Y2, X = cbind(1, dm$Y1) , K = K, verbose = F)
mg2$tau  # The genetic variance, also close to zero

lrt2 <- 2*(mg2$logL - mg2$logL0) # statistic close to zero
pchisq(lrt2, df= 1, lower.tail = F) # p-value of zero


## Third example of type (A): test if Y2 \independent G 
## Y2 ~ (1|G) , (ie S = {}) ----------------------------

## rrBLUP library ---------------------------------------------
# should give b$Vg not to zero
b3 <- kin.blup(data = dm, geno = 'G',pheno = "Y2", 
               K = K)
b3$Vg  # none zero as hoped

## lme4qtl and lme4 libraries ---------------------------------
m3 <- relmatLmer(Y2 ~ (1|G) , # Write the formula of your model
                 dm,  # Write your data table into a data frame
                 relmat = list(G = K)) # the relation matrix
vf3 <- as.data.frame(VarCorr(m3))[, c("grp", "vcov")]
vf3$prop <- with(vf3, vcov / sum(vcov))
vf3   # variance is not zero as it should

# checking Residual likelihood
RLRsim::exactRLRT(m3) #low p-value, variance is not zero

# stats::confint(m3) # intervals are .sig01[1.1,1.5]  
                   # sigma [0.8, .9] and intercept [-.09, .07]
                   # significant random effect ?


### gaston ------------------------------------------------
mg3 <- lmm.aireml(Y = dm$Y2 , K = K, verbose = F)
mg3$tau  # The genetic variance, matches above

lrt3 <- 2*(mg3$logL - mg3$logL0) # similar lrt, but not clear
pchisq(lrt3, df = 1, lower.tail = F) # how to define degrees of freedom (1 is
                     # does not give similar answer in this case

# Fourth example
## Y1 ~ (1|G) + Y2 --------------------------------------------

## rrBLUP library ---------------------------------------------
b4 <- kin.blup(data = dm, geno = 'G',pheno = "Y1", fixed = "Y2",
              K = K)
b4$Vg # model does not converge

## lme4qtl and lme4 libraries ---------------------------------
m4 <- relmatLmer(Y1 ~ (1|G) + Y2 , #  the formula of the model
                dm,  # Write your data table into a data frame
                relmat = list(G = K)) # the relation matrix

# Estimating the heritibility and checking the variances:
vf4 <- as.data.frame(VarCorr(m4))[,c("grp", "vcov")]
vf4$prop <- with(vf4, vcov / sum(vcov))
vf4 # variance is small

# RLT
RLRsim::exactRLRT(m4) # low p-value --> variance is not zero

# stats::confint(m4) # the interval: .sig01[.18, .36]
                   # .sigma [.38, .44] and intercept[-.03, .04]
                   # significant random effect


### gaston ------------------------------------------------
mg4 <- lmm.aireml(Y = dm$Y1 , X = cbind(1, dm$Y2), K = K, verbose = F)
mg4$tau  # The genetic variance, matches above

lrt4 <- 2*(mg4$logL - mg4$logL0) # congruent lrt statistic, but not
pchisq(lrt4, df = 1, lower.tail = F)             # clear how to define df


## Summary: -------------------------------------------------------------------
# In the simple network of G -> Y1 -> Y2, the expected tests and p-values should be
# as follows:
# model         | Vg  | lrt    | lme4qtl p-value  | gaston p-value| conclusion
# Y1 ~ 1|G      | !=0 | 487.57 | small            | small         | not independent
# Y2 ~ 1|G + Y1 | ==0 | ~0     | 1                | 0.5           | dependent?
# Y2 ~ 1|G      | !=0 | 347.6  | small            | small         | not independent
# Y1 ~ 1|G + Y2 | ~0  | 25.512 | small            | small         | not dependent


###############################################################
##  testing conditional independence, type (B) ############----
###############################################################

x <- dm$Y1
y <- dm$Y2
G <- dm$G

Z.t <- pcgen:::make.Z.matrix(G)
Z.t[1:6,1:6]
X.t <- Matrix(rep(1, length(G)))

em.vec <- c(x, y)
names(em.vec) <- rep(as.character(G), 2)

fit.reduced <- pcgen:::fitEM(em.vec, X.t, Z.t, cov.error = TRUE, 
                             cov.gen = FALSE, max.iter = 20)
fit.reduced2 <- fitEM(em.vec, X.t, Z.t, cov.error = TRUE, 
                      cov.gen = FALSE, max.iter = 20)
all.equal(fit.reduced, fit.reduced2)

# New code (see modification below and attached code)
source("github_repos/network-analysis/src/haplotypes/willem/EM_Function_pcgen.R")

# Both analyses give rise to the same estimates of the variance components and coefficients
fit.reduced.K <- fitEM(em.vec, X.t, Z.t, cov.error = TRUE, 
                             cov.gen = FALSE, max.iter = 20,
                             K = K)



# Changing Z defition, Much less time consuming! 
# Incorporate the idea into the algorithm (if possible)!
fit.reduced.ZM <- fitEM(em.vec, X.t, Z.t = Matrix(Z.t %*% M), 
                             cov.error = TRUE, 
                             cov.gen = FALSE, max.iter = 20)

all.equal(fit.reduced.K$coeff$fixed,fit.reduced.ZM$coeff$fixed)

qqplot(fit.reduced$coeff$random, fit.reduced.ZM$coeff$random)

str(fit.reduced.K)
str(fit.reduced.ZM)


# COTE: For the moment, we only need to make some minor modifications in this bit
# Need to explore in more detail the proposal by Marco.
if (!is.null(K)) {
        #eig <- eigen(K)
        #U <- eig$vectors
        #d <- eig$values
        ##K.trans <- U %*% diag(1/sqrt(d))
        #K.trans <- U %*% diag(sqrt(d))
        #Z.t <- Matrix(Z.t %*% K.trans)

        eig <- svd(K)
        U <- eig$u
        d <- eig$d
        #K.trans <- U %*% diag(1/sqrt(d))
        K.trans <- U %*% diag(sqrt(d))
        Z.t <- Matrix(Z.t %*% K.trans)
    }


# Illustration, to show how fitEM can work with only genotypic means:
x <- dm$Y1        
y <- dm$Y2
G <- dm$G
Z.t <- pcgen:::make.Z.matrix(G)
X.t <- Matrix(rep(1, length(G)))
em.vec <- c(x, y)
names(em.vec) <- rep(as.character(G), 2)

fit.reduced <- pcgen:::fitEM(em.vec, X.t, Z.t = Matrix(Z.t %*% M), 
                             cov.error = TRUE, 
                             cov.gen = FALSE, max.iter = 20)


```
