# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' R interface to 'Genome-wide Efficient Mixed Model Association' (GEMMA)
#'
#' GEMMA is an efficient association testing software of Linear Mixed Models
#' (LMM)s and related GWAS models. This interface is for the Multivariate
#' linear mixed model (mvLMM) functionality which also corrects for population
#' structure and sample (non)exchangability
#' @param genoinputs Character vector of input files. It should be either the
#'   prefix name of PLINK binary ped file, or the 3 BIMBAM files' names (the
#'   menan genotypes, the phenotyes and SNP annotaiton files)
#' @param kfile Character variable for the relatedness matrix file name (can be
#'   in gzip compressed format)
#' @param k_mode Integer variable for the type of the kinship/relatedness
#'   matrix type
#' @param colnums Numeric vector for specifying column number from the
#'   phenotypes file used for association testing
#' @param miss,maf,r2,hwe Floating point variables for filtering SNPs at
#'   missingness cutt-off (default 0.05), minor allele frequency (default
#'   0.01), r2 threshold (default 0.9999) and HWE test p value threshold
#'   (default 0; no test)
#' @param notsnp Boolean variable to use real values as covariates (minor
#'   allele frequency cutoff is not used)
#' @param lmmMode Integer specifying the frequenist test to use: 1 for the Wald
#'   test, 2 for the likelihood ratio test, 3 for score test; and 4 for
#'   performing all the three tests
#' @param gxe Character variable for the file containing a column of
#'   environmental variables. If provided, GEMMA fits a LMM controlling for
#'   both the SNP main effect and the environmental main effect while testing
#'   for interaction effect.
#' @param outprefix Character variable for output file prefix
#' @param outdir Character variable for output directory path
#' @param license Boolean variable for printing GEMMA's license information
#'
gemmaMVLMM <- function(genoinputs, kfile, colnums, k_mode = 1L, miss = 0.05, maf = 0.01, r2 = 0.9999, hwe = 0, notsnp = FALSE, lmmMode = 1L, gxe = "", outprefix = "result", outdir = "output", license = FALSE) {
    .Call(`_pcgen2_gemmaMVLMM`, genoinputs, kfile, colnums, k_mode, miss, maf, r2, hwe, notsnp, lmmMode, gxe, outprefix, outdir, license)
}

#' R interface to 'Genome-wide Efficient Mixed Model Association' (GEMMA)
#'
#' GEMMA is an efficient association testing software of Linear Mixed Models
#' (LMM)s and related GWAS models. This interface is for the Relatedness matrix
#' calculation functionality in centered or standardized matrix form
#' @param gk Integer variable for calculating the relatedness matrix. A value
#'   of 1 (the default) generates a centered matrix. A value of 2 generates a
#'   standardized matrix
#' @inheritParams gemmaMVLMM
#'
gemmaGK <- function(genoinputs, gk, colnums, miss = 0.05, maf = 0.01, r2 = 0.9999, hwe = 0, notsnp = FALSE, outprefix = "result", outdir = "output", license = FALSE) {
    .Call(`_pcgen2_gemmaGK`, genoinputs, gk, colnums, miss, maf, r2, hwe, notsnp, outprefix, outdir, license)
}

